app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.kt

```diff
@@ -0,0 +1,746 @@
+                    sessionConnector.setCustomErrorMessage(null)
+                    NavigationHelper.playOnBackgroundPlayer(
+                        playerService, playQueue,
+                        playWhenReady
+                    )
+                },
+                { throwable ->
+                    playbackError(
+                        ErrorInfo(
+                            throwable, UserAction.PLAY_STREAM,
+                            "Failed playback of media ID [$mediaId]: "
+                        )
+                    )
+                }
+            )
+    }
+
+    override fun onPrepareFromSearch(
+        query: String,
+        playWhenReady: Boolean,
+        extras: Bundle?
+    ) {
+        disposePrepareOrPlayCommands()
+        playbackError(
+            R.string.content_not_supported,
+            PlaybackStateCompat.ERROR_CODE_NOT_SUPPORTED
+        )
+    }
+
+    private fun searchMusicBySongTitle(query: String): Single<SearchInfo> {
+        val serviceId = ServiceHelper.getSelectedServiceId(playerService)
+        return ExtractorHelper.searchFor(
+            serviceId, query,
+            ArrayList(), ""
+        )
+    }
+
+    private fun mediaItemsFromInfoItemList(result: ListInfo<InfoItem>): SingleSource<List<MediaBrowserCompat.MediaItem>> {
+        result.errors
+            .takeIf { exceptions ->
+                exceptions.isNotEmpty() &&
+                    !(
+                        exceptions.size == 1 &&
+                            exceptions.first() is NothingFoundException
+                        )
+            }?.let { exceptions ->
+                return Single.error(exceptions.first())
+            }
+
+        val items = result.relatedItems
+        if (items.isEmpty()) {
+            return Single.error(NullPointerException("Got no search results."))
+        }
+        try {
+            val results =
+                items
+                    .filter { item ->
+                        item.infoType == InfoType.STREAM ||
+                            item.infoType == InfoType.PLAYLIST ||
+                            item.infoType == InfoType.CHANNEL
+                    }.map { item -> this.createInfoItemMediaItem(item) }
+
+            return Single.just(results)
+        } catch (error: Exception) {
+            return Single.error(error)
+        }
+    }
+
+    private fun handleSearchError(throwable: Throwable) {
+        Log.e(TAG, "Search error: $throwable")
+        disposePrepareOrPlayCommands()
+        sessionConnector.setCustomErrorMessage(
+            playerService.getString(R.string.search_no_results),
+            PlaybackStateCompat.ERROR_CODE_APP_ERROR,
+        )
+    }
+
+    override fun onPrepareFromUri(
+        uri: Uri,
+        playWhenReady: Boolean,
+        extras: Bundle?
+    ) {
+        disposePrepareOrPlayCommands()
+        playbackError(R.string.content_not_supported, PlaybackStateCompat.ERROR_CODE_NOT_SUPPORTED)
```

Comment:
What are `onPrepareFromSearch` and `onPrepareFromUri` supposed to do?

Reply:
As far as I understand, Google Assistant uses these APIs when it acts as a controller of a media application (e.g., following a command like "play <search query> music on NewPipe". Unfortunately, I believe only Play Store apps can be controlled by Google Assistant, so there wasn't much point in implementing them.

Reply:
This might be a niche case, but you can set the app as "installed from PlayStore" by using KingInstalled or from ADB. To my knowledge this however currently only works upto Android 13, so might not be worth the effort.

Reply:
No it's not worth the effort (shame on Google)